### PR TITLE
USHIFT-1901: Retry once failed composer build

### PR DESCRIPTION
**Pull request by pmtk**
Adds a retry when a build job fails. It can be tested with `composer-cli compose cancel` is it marks the job as FAILED.

### CODE REVIEW

**Comment by pmtk** [comment]
At least we know it works :)
```
2023-11-23 22:10:41,043: Waiting for ['96d68ead-d408-45bb-b373-7393c92dee0a', 'ab3b0cb0-a59f-4e26-911a-c633bdf78140', 'd72d30ff-0014-46f5-a966-e8fc2479bfe8', 'd6a1c807-91a7-43c4-aff1-3fd99c14943f', 'a2d46b39-73a3-4d48-824c-735183892065']
2023-11-23 22:10:41,135: a2d46b39-73a3-4d48-824c-735183892065 edge-commit for rhel-9.2-microshift-source - RUNNING
2023-11-23 22:10:41,135: 96d68ead-d408-45bb-b373-7393c92dee0a edge-commit for rhel-9.2-microshift-crel - RUNNING
2023-11-23 22:10:41,135: ab3b0cb0-a59f-4e26-911a-c633bdf78140 edge-commit for rhel-9.2-microshift-source-base - RUNNING
2023-11-23 22:10:41,135: d72d30ff-0014-46f5-a966-e8fc2479bfe8 edge-commit for rhel-9.2-microshift-source-fake-next-minor - RUNNING
2023-11-23 22:10:41,135: d6a1c807-91a7-43c4-aff1-3fd99c14943f edge-commit for rhel-9.2-microshift-source-fake-yplus2-minor - RUNNING
2023-11-23 22:11:11,165: Waiting for ['96d68ead-d408-45bb-b373-7393c92dee0a', 'ab3b0cb0-a59f-4e26-911a-c633bdf78140', 'd72d30ff-0014-46f5-a966-e8fc2479bfe8', 'd6a1c807-91a7-43c4-aff1-3fd99c14943f', 'a2d46b39-73a3-4d48-824c-735183892065']
2023-11-23 22:11:11,256: 96d68ead-d408-45bb-b373-7393c92dee0a edge-commit for rhel-9.2-microshift-crel - FAILED
2023-11-23 22:11:11,256: Job 96d68ead-d408-45bb-b373-7393c92dee0a failed - restarting once (sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://10.192.10.201:8080/repo --ref rhel-9.2-microshift-crel rhel-9.2-microshift-crel edge-commit)
2023-11-23 22:11:13,455: Job 96d68ead-d408-45bb-b373-7393c92dee0a restarted as 704180b5-6fee-40f2-9952-7019c0417c8d
2023-11-23 22:11:13,455: a2d46b39-73a3-4d48-824c-735183892065 edge-commit for rhel-9.2-microshift-source - FAILED
2023-11-23 22:11:13,455: Job a2d46b39-73a3-4d48-824c-735183892065 failed - restarting once (sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://10.192.10.201:8080/repo --ref rhel-9.2-microshift-source rhel-9.2-microshift-source edge-commit)
2023-11-23 22:11:14,554: Job a2d46b39-73a3-4d48-824c-735183892065 restarted as 610068f5-22fd-4648-a990-267ec1f9aee8
2023-11-23 22:11:14,555: ab3b0cb0-a59f-4e26-911a-c633bdf78140 edge-commit for rhel-9.2-microshift-source-base - FAILED
2023-11-23 22:11:14,555: Job ab3b0cb0-a59f-4e26-911a-c633bdf78140 failed - restarting once (sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://10.192.10.201:8080/repo --ref rhel-9.2-microshift-source-base rhel-9.2-microshift-source-base edge-commit)
2023-11-23 22:11:15,636: Job ab3b0cb0-a59f-4e26-911a-c633bdf78140 restarted as 1a64da7e-ac07-45ae-8b55-2b9b021ff7b2
2023-11-23 22:11:15,636: d6a1c807-91a7-43c4-aff1-3fd99c14943f edge-commit for rhel-9.2-microshift-source-fake-yplus2-minor - FAILED
2023-11-23 22:11:15,636: Job d6a1c807-91a7-43c4-aff1-3fd99c14943f failed - restarting once (sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://10.192.10.201:8080/repo --ref rhel-9.2-microshift-source-fake-yplus2-minor rhel-9.2-microshift-source-fake-yplus2-minor edge-commit)
2023-11-23 22:11:16,731: Job d6a1c807-91a7-43c4-aff1-3fd99c14943f restarted as d51d2e20-8e93-4af3-b6b4-0713ab13f152
2023-11-23 22:11:16,731: d72d30ff-0014-46f5-a966-e8fc2479bfe8 edge-commit for rhel-9.2-microshift-source-fake-next-minor - FAILED
2023-11-23 22:11:16,731: Job d72d30ff-0014-46f5-a966-e8fc2479bfe8 failed - restarting once (sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://10.192.10.201:8080/repo --ref rhel-9.2-microshift-source-fake-next-minor rhel-9.2-microshift-source-fake-next-minor edge-commit)
2023-11-23 22:11:17,823: Job d72d30ff-0014-46f5-a966-e8fc2479bfe8 restarted as 146b0fbb-9b9c-460e-a5dc-e948dd4ea6cb
2023-11-23 22:11:47,854: Waiting for ['704180b5-6fee-40f2-9952-7019c0417c8d', '610068f5-22fd-4648-a990-267ec1f9aee8', '1a64da7e-ac07-45ae-8b55-2b9b021ff7b2', 'd51d2e20-8e93-4af3-b6b4-0713ab13f152', '146b0fbb-9b9c-460e-a5dc-e948dd4ea6cb']
2023-11-23 22:11:47,974: 146b0fbb-9b9c-460e-a5dc-e948dd4ea6cb edge-commit for rhel-9.2-microshift-source-fake-next-minor - FAILED
2023-11-23 22:11:47,974: Job 146b0fbb-9b9c-460e-a5dc-e948dd4ea6cb failed - not restarting anymore (empty cmd)
2023-11-23 22:11:47,974: 1a64da7e-ac07-45ae-8b55-2b9b021ff7b2 edge-commit for rhel-9.2-microshift-source-base - FAILED
2023-11-23 22:11:47,974: Job 1a64da7e-ac07-45ae-8b55-2b9b021ff7b2 failed - not restarting anymore (empty cmd)
2023-11-23 22:11:47,974: 610068f5-22fd-4648-a990-267ec1f9aee8 edge-commit for rhel-9.2-microshift-source - FAILED
2023-11-23 22:11:47,974: Job 610068f5-22fd-4648-a990-267ec1f9aee8 failed - not restarting anymore (empty cmd)
2023-11-23 22:11:47,974: 704180b5-6fee-40f2-9952-7019c0417c8d edge-commit for rhel-9.2-microshift-crel - FAILED
2023-11-23 22:11:47,974: Job 704180b5-6fee-40f2-9952-7019c0417c8d failed - not restarting anymore (empty cmd)
2023-11-23 22:11:47,974: d51d2e20-8e93-4af3-b6b4-0713ab13f152 edge-commit for rhel-9.2-microshift-source-fake-yplus2-minor - FAILED
2023-11-23 22:11:47,974: Job d51d2e20-8e93-4af3-b6b4-0713ab13f152 failed - not restarting anymore (empty cmd)
```

File: test/bin/build_images.sh [hunk 1]
```diff
@@ -270,47 +271,43 @@ do_group() {
...
-                          | awk '{print $2}')
+            build_cmd="sudo composer-cli compose start ${blueprint} image-installer"
+            buildid=$(${build_cmd} | awk '{print $2}')
             echo "Build ID ${buildid}"
             # Record a "build name" to be used as part of the unique
             # filename for the log we download next.
             echo "${blueprint}-image-installer" > "${IMAGEDIR}/builds/${buildid}.build"
-            buildid_list="${buildid_list} ${buildid}"
+            buildid_list+=("${buildid},${build_cmd}")
+            #buildid_list="${buildid_list}#\"${buildid},${build_cmd}\""
```
Comment (ggiguash) [inline test/bin/build_images.sh:294]: Remove comment?

File: test/bin/build_images.sh [hunk 2]
```diff
@@ -270,47 +271,43 @@ do_group() {
...
     fi
 
-    if [ -n "${buildid_list}" ]; then
+    if (( ${#buildid_list[@]} )); then
         echo "Waiting for builds to complete..."
-        # shellcheck disable=SC2086  # pass command arguments quotes to allow word splitting
-        time "${SCRIPTDIR}/wait_images.py" ${buildid_list}
+        # wait_images.py returns possibly updated list of builds that must be handled
+        # "update" means replacing initial build ID with retry build ID
+        builds_to_get=$(time "${SCRIPTDIR}/wait_images.py" "${buildid_list[@]}")
```
Comment (ggiguash) [inline test/bin/build_images.sh:302]: Should we better save those builds in a file? Transferring them by stdout seems a bit dangerous if a script fails, etc.

Comment (ggiguash) [reply test/bin/build_images.sh:302]: File name can be an argument to the wait_images.py script.

File: test/bin/build_images.sh [hunk 3]
```diff
@@ -197,7 +197,8 @@ do_group() {
     local blueprint_file
     local build_name
     local buildid
-    local buildid_list=""
+    local buildid_list=()
+    local builds_to_get=""
```
Comment (ggiguash) [inline test/bin/build_images.sh:201]: Nit: this declaration can be moved down to the scope where it's used.

Comment (pmtk) [reply test/bin/build_images.sh:201]: Yeah, but it seems that we went with declaring them at the top of function so didn't want to diverge from the style

File: test/bin/build_images.sh (hunk 2 above)
Comment (pmtk) [reply test/bin/build_images.sh:302]: Do you think the risk is significant that we need to complicate the data flow?

To test I added `sys.exit("test")` before `main(input)`:

```
+ echo 'Waiting for builds to complete...'
Waiting for builds to complete...
++ /home/pm/microshift/test/bin/wait_images.py '487bd1da-33d3-4b0f-9664-8e9ce00afbfb,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-crel rhel-9.2-microshift-crel edge-commit' 'b8a17c36-5b32-4c50-9cb2-8f56bf8af9f8,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-base rhel-9.2-microshift-source-base edge-commit' 'd1e32594-cc99-4250-8c59-bc70c527b304,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-fake-next-minor rhel-9.2-microshift-source-fake-next-minor edge-commit' 'b398a5de-f0c6-42a4-a261-8d7da2185293,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-fake-yplus2-minor rhel-9.2-microshift-source-fake-yplus2-minor edge-commit' '8bd6b8d5-05e6-4655-8dca-05eefe57d527,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source rhel-9.2-microshift-source edge-commit'
2023-11-24 10:44:15,209: Received arguments: {'487bd1da-33d3-4b0f-9664-8e9ce00afbfb': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-crel rhel-9.2-microshift-crel edge-commit', 'b8a17c36-5b32-4c50-9cb2-8f56bf8af9f8': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-base rhel-9.2-microshift-source-base edge-commit', 'd1e32594-cc99-4250-8c59-bc70c527b304': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-fake-next-minor rhel-9.2-microshift-source-fake-next-minor edge-commit', 'b398a5de-f0c6-42a4-a261-8d7da2185293': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-fake-yplus2-minor rhel-9.2-microshift-source-fake-yplus2-minor edge-commit', '8bd6b8d5-05e6-4655-8dca-05eefe57d527': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source rhel-9.2-microshift-source edge-commit'}
test

real    0m0.031s
user    0m0.027s
sys     0m0.004s
+ builds_to_get=
+ osbuild_logs
++ sudo systemctl list-units
++ awk '/osbuild-worker@/ {print $1} /osbuild-composer\.service/ {print $1}'
+ workers_services='osbuild-composer.service
osbuild-worker@1.service
osbuild-worker@2.service
osbuild-worker@3.service
osbuild-worker@4.service
osbuild-worker@5.service'
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-composer.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@1.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@2.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@3.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@4.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@5.service

$ echo $?
1
```

Comment (ggiguash) [reply test/bin/build_images.sh:302]: OK, so if the script exits with error, the procedure fails - that's good.
I'm not sure if we absolutely need to implement this. I'm just suspicious of program output usage for follow-up actions in general.
Let's decide about it on Monday

Comment (pmtk) [reply test/bin/build_images.sh:302]: But we do it all the time, right? Even for starting the jobs we take the UUID from the output.

**Comment by dhellmann** [comment]
/lgtm

We agreed today to merge this as-is and do a more substantial rewrite of the image build stage as a separate task.

**Comment by ggiguash** [comment]
Test does not relate to the changes
/override ci/prow/metal-periodic-test

**Comment by pmtk** [comment]
/override ci/prow/metal-periodic-test ci/prow/microshift-metal-tests ci/prow/ocp-conformance-rhel-eus-arm

**Pull request merged**